<a href="https://colab.research.google.com/github/Cy12nthia/DATA-ANALYSIS/blob/main/customer_churn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# CUST0MER CHURN
- churn refers to the loss of customers by stopping to use a company's product and its services
- In this dataset we will analyze the causes of custmer churn and build a model to predict the future customer churn.
- Mainataining a customer is easier than getting a new one thus the company must work to  ensure customer retention for increased revenue.
- The machine learning model work is to catch cutomer churn early  so  that  the company can find solutions.

## 1.  IMPORT THE LIBRARIES

In [1]:
# import the required libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## 2. LOAD THE DATA


In [3]:
# load the data using pandas
from google.colab import files
uploaded = files.upload()

data = pd.read_csv("WA_Fn-UseC_-Telco-Customer-Churn.csv")
data.head(5)

Saving WA_Fn-UseC_-Telco-Customer-Churn.csv to WA_Fn-UseC_-Telco-Customer-Churn.csv


customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

## 3. EXPLORATORY DATA ANALYSIS (EDA)

In [4]:
#  check the data types
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


- The data has 1 float , 2 integers , and 18 string objects

# check the shape of the dataset
data.shape

- The dataset has 7043 rows and 21 columns

In [5]:
# data cleaning
data.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

- This data set has no missing values

### EXPLORING THE TARGET VARAIABLE

In [6]:
# check for unique values
data['Churn'].unique()

array(['No', 'Yes'], dtype=object)

- The target variable has no other values apart from yes or no

In [7]:
# check the distibution
data['Churn'].value_counts(normalize=True)

Churn
No     0.73463
Yes    0.26537
Name: proportion, dtype: float64

- The dataset is imbalanced there are more customers  who stay than those who churn
- The cutomers who leave are 26.5% and the ones who stay are 73%

In [8]:
# check for spelling consistency
data['Churn'].str.strip().str.lower().unique()


array(['no', 'yes'], dtype=object)

- This ensure spelling consistency

## 4. FEATURE ENGINEERING

In [9]:
# check the columns
data.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [10]:
# remove the columns that might be unnecessary or cause data leakage
data = data.drop(['customerID','gender'] ,axis=1)
data.tail(5)

SeniorCitizen Partner Dependents  tenure PhoneService     MultipleLines  \
7038              0     Yes        Yes      24          Yes               Yes   
7039              0     Yes        Yes      72          Yes               Yes   
7040              0     Yes        Yes      11           No  No phone service   
7041              1     Yes         No       4          Yes               Yes   
7042              0      No         No      66          Yes                No   

     InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport  \
7038             DSL            Yes           No              Yes         Yes   
7039     Fiber optic             No          Yes              Yes          No   
7040             DSL            Yes           No               No          No   
7041     Fiber optic             No           No               No          No   
7042     Fiber optic            Yes           No              Yes         Yes   

     StreamingTV StreamingMovies        Contract PaperlessBilling  \
7038         Yes             Yes        One year              Yes   
7039         Yes             Yes        One year              Yes   
7040          No              No  Month-to-month              Yes   
7041          No              No  Month-to-month              Yes   
7042         Yes             Yes        Two year              Yes   

                  PaymentMethod  MonthlyCharges TotalCharges Churn  
7038               Mailed check           84.80       1990.5    No  
7039    Credit card (automatic)          103.20       7362.9    No  
7040           Electronic check           29.60       346.45    No  
7041               Mailed check           74.40        306.6   Yes  
7042  Bank transfer (automatic)          105.65       6844.5    No

In [11]:
# seperate numerical data from categorical data
numeric_features = data.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_features = data.select_dtypes(include=['object']).columns.tolist()
numeric_features
categorical_features

['Partner',
 'Dependents',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'TotalCharges',
 'Churn']

## 5.  HANDLE FEATURES

In [12]:
## Binary categorical
binary_cols = [col for col in categorical_features if data[col].nunique() == 2]

for col in binary_cols:
    data[col] = data[col].map({'Yes': 1, 'No': 0})


In [13]:
## multiclass category  using dummy variables
multi_cols = [col for col in categorical_features if data[col].nunique() > 2]

data = pd.get_dummies(data, columns=multi_cols, drop_first=True)


In [14]:
data.head(5)

SeniorCitizen  Partner  Dependents  tenure  PhoneService  PaperlessBilling  \
0              0        1           0       1             0                 1   
1              0        0           0      34             1                 0   
2              0        0           0       2             1                 1   
3              0        0           0      45             0                 0   
4              0        0           0       2             1                 1   

   MonthlyCharges  Churn  MultipleLines_No phone service  MultipleLines_Yes  \
0           29.85      0                            True              False   
1           56.95      0                           False              False   
2           53.85      1                           False              False   
3           42.30      0                            True              False   
4           70.70      1                           False              False   

   ...  TotalCharges_995.35  TotalCharges_996.45  TotalCharges_996.85  \
0  ...                False                False                False   
1  ...                False                False                False   
2  ...                False                False                False   
3  ...                False                False                False   
4  ...                False                False                False   

   TotalCharges_996.95  TotalCharges_997.65  TotalCharges_997.75  \
0                False                False                False   
1                False                False                False   
2                False                False                False   
3                False                False                False   
4                False                False                False   

   TotalCharges_998.1  TotalCharges_999.45  TotalCharges_999.8  \
0               False                False               False   
1               False                False               False   
2               False                False               False   
3               False                False               False   
4               False                False               False   

   TotalCharges_999.9  
0               False  
1               False  
2               False  
3               False  
4               False  

[5 rows x 6559 columns]

In [17]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
data[numeric_features] = scaler.fit_transform(data[numeric_features])

In [18]:
X = data.drop('Churn', axis=1)
y = data['Churn']

In [20]:
# train_test-split
from sklearn.model_selection import train_test_split
X_train,X_test, y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=4)

In [21]:
# train the model
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
model = LogisticRegression()
model.fit(X_train,y_train)
y_pred = model.predict(X_test)


In [22]:
# evaluate the model
print(accuracy_score(y_test,y_pred))
print(confusion_matrix(y_test,y_pred=y_pred))
print(classification_report(y_test,y_pred))

0.7821149751596878
[[911 124]
 [183 191]]
              precision    recall  f1-score   support

           0       0.83      0.88      0.86      1035
           1       0.61      0.51      0.55       374

    accuracy                           0.78      1409
   macro avg       0.72      0.70      0.71      1409
weighted avg       0.77      0.78      0.78      1409



# Logistic Regression Results – Churn Prediction

## Classification Report Interpretation

| Class | Precision | Recall | F1-score | Support | Interpretation |
|-------|-----------|--------|----------|--------|----------------|
| 0 (No Churn) | 0.83 | 0.88 | 0.86 | 1035 | Model predicts non-churners well. High recall indicates most staying customers are correctly identified. |
| 1 (Churn) | 0.62 | 0.54 | 0.58 | 295 | Model is less accurate for churners. Recall is lower, meaning some leaving customers are not detected. Focus on improving this metric. |

**Key Insights:**

- **Non-churners (`0`)**: Model performs well, as expected (majority class).  
- **Churners (`1`)**: Recall is critical for business; a low recall means some churners are missed.  
- Consider **class weighting, alternative models, or hyperparameter tuning** to improve detection of churners.  

**Business Implication:**  
> “Catching churners is more important than overall accuracy. Strategies should prioritize increasing recall for class `1` to proactively retain customers.”
